// uni-simple-router/helpers/config.js
export const baseConfig={
	h5:{
    loading:true,		//是否显示加载动画
    hinderTab:false,  //是否拦截uni-app自带底部菜单   TODO
		vueRouterDev:false,	//完全使用采用vue-router的开发模式
    useUniConfig:true,	//是否采用在pages.json下的所有页面配置信息,false时需开发者自行设置页面
    keepUniIntercept:false,   //保留uni-app使用vue-router的拦截器
    vueNext:false,  //在next管道函数中是否获取vueRouter next的原本参数
		mode:'hash',
		base:'/',
		linkActiveClass:'router-link-active',
		linkExactActiveClass:'router-link-exact-active',
		scrollBehavior:(to, from, savedPostion)=>savedPostion,
    fallback:true,
  },
  encodeURI:true,   //是否对url传递的参数进行编码
	routes:[],
}
export const methods = {
  push: "navigateTo",
  replace: "redirectTo",
  replaceAll: "reLaunch",
  pushTab: "switchTab",
  back: "navigateBack"
};
export const lifeCycle = {
  beforeHooks: [],
  afterHooks: [],
  routerHooks: [],
  routerbeforeHooks:[],		//内部跳转前生命周期
  routerAfterHooks:[],	//内部跳转后生命周期
};

export const Global={ //缓存一些必要的对象，作为全局可以访问的参数
  Router:{},
}

export const route = function(object = {}) {
  return {
    ...object,
    params: {},
    query: {}
  };
};
