// uni-simple-router/lifeCycle/hooks.js
import {
	resolveRule,
	normalizeParams,
	parseQuery,
	queryMp,
	appPlatform
} from "../helpers/util.js";

import {
	route
} from "../helpers/config.js";

import {
	queryInfo
} from "../patch/applets-patch.js";

import {
	err
} from "../helpers/warn.js";

const platform = appPlatform(true);

export const registerHook = function (list, fn) {
	list.push(fn);
	return () => {
		const i = list.indexOf(fn);
		if (i > -1) list.splice(i, 1);
	};
};
export const executeHook = function (list, ...args) {
	for (let i = 0; i < list.length; i++) {
		list[i](args);
	}
};

export const isNext = function (router, Intercept, rule, fnType) {
	return new Promise(async (resolve, reject) => {
		if (Intercept === false || Intercept === 0) {
			return reject("路由终止");
		}

		if (Intercept === true || Intercept === undefined) {
			return resolve({
				toRule: rule,
				ags: {
					rule: {
						NAVTYPE: fnType
					}
				}
			});
		}
		if (Intercept.constructor === String) {
			Intercept = {
				path: Intercept,
				NAVTYPE: fnType
			}
		} else {
			if (Reflect.get(Intercept, 'NAVTYPE') === undefined) {
				Intercept.NAVTYPE = fnType;
			}
		}
		const nextRule=await resolveParams(router, Intercept, fnType);
		return resolve(nextRule);
	});
};

export const resolveHooks=function({
	ags,
	_to,
	navigateFun,
	router,
	fnType,
}){
	return new Promise(async resolve=>{
		let fromatRule={};
		const beforeResult = await resolveBeforeHooks(ags);
		if (navigateFun == null) {
			return resolve({
				toRule: _to,
				beforeResult,
				ags
			});
		}
		fromatRule = await isNext(router, beforeResult, _to, fnType);
	
		fnType = fromatRule.ags.rule.NAVTYPE;
	
		const beforeEnter = Reflect.get(fromatRule.toRule.rule, "beforeEnter");
		if (beforeEnter) {
			fromatRule = await resolveRouterHooks(
				ags,
				fnType,
				beforeEnter,
				fromatRule.toRule,
				navigateFun
			);
		}
		resolve(fromatRule);
	})

}

export const resolveParams = async function (router, rule, fnType, navigateFun) {
	if (navigateFun != null) {
		const isComponent = await router.lifeCycle["routerbeforeHooks"][0].call(router, fnType) //触发内部跳转前的生命周期,并验证当前是否为组件
		if (!isComponent) {
			return err('Vue模板未编译完成，不支持跳转。请检查 $Router API 代码')
		}
	}
	
	router.lastVim = queryMp(router.lastVim);
	const routeInfo = queryInfo(router.lastVim);
	const _from = resolveRule(router, routeInfo.route, routeInfo.query);
	const _to = normalizeParams(JSON.parse(JSON.stringify(rule)), router.CONFIG.routes);
	const ags = {
		router,
		fromRule: _from,
		toRule: {
			..._to.rule,
			...route(),
			query: parseQuery("query", _to.query, true).query
		},
		rule,
		fnType
	};
	let navFunCallback = false;
	try {
		const fromatRule= await resolveHooks({
			_to,
			ags,
			router,
			navigateFun,
			fnType
		});
		if(navigateFun==null){
			return fromatRule;
		}
		navFunCallback = await navigateFun.call(router, fromatRule);
		resolveAfterHooks(
			router, {
				...route(),
				...fromatRule.toRule.rule,
				query: parseQuery("query", fromatRule.toRule.query, true).query
			},
			ags.fromRule
		);
	} catch (e) {};

	if (navigateFun != null) {
		router.lifeCycle["routerAfterHooks"][0].call(router, navFunCallback) //触发内部跳转后的生命周期
	}

};
/**
 * 触发全局afterHooks 生命钩子
 */
export const resolveAfterHooks = function (router, toRule, fromRule) {
	if (platform == 'H5') {
		return Promise.resolve();
	}
	return new Promise(async (resolve, rejcet) => {
		if (!router.lifeCycle["afterHooks"][0]) {
			return resolve();
		}
		await router.lifeCycle["afterHooks"][0](toRule, fromRule, resolve);
	});
};

/**
 * 触发全局beforeHooks 生命钩子
 */
export const resolveBeforeHooks = function ({
	router,
	fromRule,
	toRule,
	rule,
	fnType
} = {}) {
	if (platform == 'H5') {
		return Promise.resolve();
	}
	return new Promise(async (resolve, reject) => {
		if (!router.lifeCycle["beforeHooks"][0]) {
			return resolve();
		}
		await router.lifeCycle["beforeHooks"][0](toRule, fromRule, resolve);
	});
};
/**
 * 触发路由独享的守卫 beforeEnter 生命钩子
 */
export const resolveRouterHooks = function ({
		router,
		fromRule
	} = {},
	fnType,
	hookFun,
	rule,
	navigateFun
) {
	return new Promise(async (resolve, reject) => {
		const Intercept = await new Promise(async resolve => {
			await hookFun({
					...route(),
					...rule.rule,
					query: parseQuery("query", rule.query, true).query
				},
				fromRule,
				resolve
			);
		});
		let fromatRule = await isNext(router, Intercept, rule, fnType);
		const nextAgs=fromatRule.ags;
		const {beforeResult}=fromatRule;
		if(beforeResult){		//如果全局前置守卫有返回数据 直接执行全局守卫
			const {NAVTYPE}=beforeResult;
			delete beforeResult.NAVTYPE;
			fromatRule = await isNext(router, beforeResult, rule, NAVTYPE);
		}else if (Object.keys(nextAgs).length > 0) {		
			const beforeEnter = Reflect.get(nextAgs.toRule || {}, "beforeEnter");
			if (beforeEnter) {
				return resolve(
					await resolveRouterHooks(
						nextAgs,
						fnType,
						beforeEnter,
						fromatRule.toRule,
						navigateFun
					)
				);
			}
		}
		resolve(fromatRule);
	});
};
